File: agents/adapters/llm/src/llm_adapter/providers/ollama.py
# SPDX-License-Identifier: Apache-2.0
"""Ollama provider — streams token chunks via httpx.AsyncClient REST API."""

from __future__ import annotations

import json
from collections.abc import AsyncGenerator
from typing import TYPE_CHECKING

import httpx
import structlog

if TYPE_CHECKING:
    from llm_adapter.config import OllamaProviderConfig

log = structlog.get_logger()

_CHAT_PATH = "/api/chat"


async def stream_tokens(
    config: OllamaProviderConfig,
    prompt: str,
    system: str | None,
    temperature: float,
    max_tokens: int,
) -> AsyncGenerator[bytes, None]:
    """Stream token chunks from the Ollama REST API.

    Args:
        config: Ollama provider configuration (base_url + model).
        prompt: User prompt.
        system: Optional system message.
        temperature: Sampling temperature.
        max_tokens: Maximum token ceiling (mapped to ``num_predict``).

    Yields:
        UTF-8 encoded token chunk bytes.

    Raises:
        RuntimeError: On HTTP or connection error; mapped to UPSTREAM_ERROR by the handler.
    """
    messages = []
    if system:
        messages.append({"role": "system", "content": system})
    messages.append({"role": "user", "content": prompt})

    body = {
        "model": config.model,
        "messages": messages,
        "stream": True,
        "options": {"temperature": temperature, "num_predict": max_tokens},
    }
    url = config.base_url.rstrip("/") + _CHAT_PATH

    timeout = httpx.Timeout(connect=5.0, read=120.0, write=10.0, pool=5.0)
    try:
        async with httpx.AsyncClient(timeout=timeout) as client:
            async with client.stream("POST", url, json=body) as response:
                response.raise_for_status()
                async for line in response.aiter_lines():
                    if not line:
                        continue
                    data = json.loads(line)
                    content = data.get("message", {}).get("content", "")
                    if content:
                        yield content.encode()
                    if data.get("done"):
                        break
    except httpx.HTTPError as exc:
        log.warning("ollama_upstream_error", error=str(exc)[:256])
        raise RuntimeError(str(exc)[:512]) from exc

File: agents/adapters/llm/internal/provider/ollama.go
// SPDX-License-Identifier: Apache-2.0

package provider

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/zynax-io/zynax/agents/adapters/llm/internal/config"
)

// ollamaChatPath is the Ollama REST streaming chat endpoint.
const ollamaChatPath = "/api/chat"

// ollamaProvider streams token chunks from an Ollama server over net/http using
// the /api/chat newline-delimited-JSON streaming protocol. It needs no
// credential. The http.Client is injectable so tests drive it with httptest.
type ollamaProvider struct {
	model     string
	baseURL   string
	maxTokens int
	client    *http.Client
}

// newOllamaProvider binds Ollama config into a Provider. A nil-free default
// client with a generous read budget is used; tests replace client directly.
func newOllamaProvider(cfg config.ProviderConfig) *ollamaProvider {
	return &ollamaProvider{
		model:     cfg.Model,
		baseURL:   strings.TrimRight(cfg.OllamaBaseURL, "/"),
		maxTokens: effectiveMaxTokens(cfg),
		client:    &http.Client{Timeout: 120 * time.Second},
	}
}

// ollamaMessage is one role/content turn in the chat request.
type ollamaMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// ollamaRequest is the JSON body posted to /api/chat.
type ollamaRequest struct {
	Model    string          `json:"model"`
	Messages []ollamaMessage `json:"messages"`
	Stream   bool            `json:"stream"`
	Options  ollamaOptions   `json:"options"`
}

// ollamaOptions carries inference parameters Ollama accepts.
type ollamaOptions struct {
	NumPredict int `json:"num_predict"`
}

// ollamaStreamLine is one NDJSON frame from the streaming response.
type ollamaStreamLine struct {
	Message struct {
		Content string `json:"content"`
	} `json:"message"`
	Done bool `json:"done"`
}

// Stream posts the prompt and relays each NDJSON message.content as a Chunk.
func (p *ollamaProvider) Stream(ctx context.Context, prompt string) (<-chan Chunk, error) {
	body, err := json.Marshal(ollamaRequest{
		Model:    p.model,
		Messages: []ollamaMessage{{Role: "user", Content: prompt}},
		Stream:   true,
		Options:  ollamaOptions{NumPredict: p.maxTokens},
	})
	if err != nil {
		return nil, fmt.Errorf("ollama: marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, p.baseURL+ollamaChatPath, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("ollama: build request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	out := make(chan Chunk)
	go p.run(req, out)
	return out, nil
}

// run executes the request and pumps the NDJSON body into out, closing it when
// the stream ends or fails. It never leaks credentials or raw bodies.
func (p *ollamaProvider) run(req *http.Request, out chan<- Chunk) {
	defer close(out)

	resp, err := p.client.Do(req) //nolint:gosec // URL host is operator-controlled config (ollama_base_url), not request input
	if err != nil {
		sendErr(out, fmt.Errorf("ollama: %s", sanitiseErr(err.Error())))
		return
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		sendErr(out, fmt.Errorf("ollama: upstream returned HTTP %d", resp.StatusCode))
		return
	}

	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	for scanner.Scan() {
		line := bytes.TrimSpace(scanner.Bytes())
		if len(line) == 0 {
			continue
		}
		var frame ollamaStreamLine
		if err := json.Unmarshal(line, &frame); err != nil {
			sendErr(out, fmt.Errorf("ollama: decode stream frame: %s", sanitiseErr(err.Error())))
			return
		}
		if frame.Message.Content != "" {
			if !send(req.Context(), out, Chunk{Text: frame.Message.Content}) {
				return
			}
		}
		if frame.Done {
			return
		}
	}
	if err := scanner.Err(); err != nil {
		sendErr(out, fmt.Errorf("ollama: read stream: %s", sanitiseErr(err.Error())))
	}
}
